File: dsa-problems/GameTheory/solutions/cpp/GMT-011-blocking-tokens-dag.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    vector<vector<int>> adj;
    vector<vector<int>> memo; // 0: unknown, 1: losing, 2: winning

    bool canWin(int u, int v) {
        if (memo[u][v] != 0) return memo[u][v] == 2;

        bool canReachLosing = false;

        // Try moving u
        for (int nextU : adj[u]) {
            if (nextU == v) continue;
            if (!canWin(nextU, v)) {
                canReachLosing = true;
                break;
            }
        }

        // Try moving v
        if (!canReachLosing) {
            for (int nextV : adj[v]) {
                if (nextV == u) continue;
                if (!canWin(u, nextV)) {
                    canReachLosing = true;
                    break;
                }
            }
        }

        memo[u][v] = canReachLosing ? 2 : 1;
        return canReachLosing;
    }

public:
    string blockingTokens(int n, vector<vector<int>>& edges, int u, int v) {
        adj.assign(n + 1, vector<int>());
        for (const auto& e : edges) {
            adj[e[0]].push_back(e[1]);
        }

        memo.assign(n + 1, vector<int>(n + 1, 0));
        return canWin(u, v) ? "First" : "Second";
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n, m;
    if (cin >> n >> m) {
        vector<vector<int>> edges(m, vector<int>(2));
        for (int i = 0; i < m; i++) {
            cin >> edges[i][0] >> edges[i][1];
        }
        int u, v;
        cin >> u >> v;
        
        Solution solution;
        cout << solution.blockingTokens(n, edges, u, v) << "\n";
    }
    return 0;
}

File: dsa-problems/GameTheory/solutions/python/GMT-011-blocking-tokens-dag.py
from typing import List
import sys

# Increase recursion depth just in case
sys.setrecursionlimit(20000)

def blocking_tokens(n: int, edges: List[List[int]], u: int, v: int) -> str:
    adj = [[] for _ in range(n + 1)]
    for a, b in edges:
        adj[a].append(b)
    
    # memo[u][v] stores result: None=unknown, False=Losing, True=Winning
    memo = {}

    def can_win(curr_u, curr_v):
        state = (curr_u, curr_v)
        if state in memo:
            return memo[state]
        
        # Try moving u
        for next_u in adj[curr_u]:
            if next_u == curr_v: continue
            if not can_win(next_u, curr_v):
                memo[state] = True
                return True
        
        # Try moving v
        for next_v in adj[curr_v]:
            if next_v == curr_u: continue
            if not can_win(curr_u, next_v):
                memo[state] = True
                return True
                
        memo[state] = False
        return False

    return "First" if can_win(u, v) else "Second"

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        m = int(next(iterator))
        edges = []
        for _ in range(m):
            u_edge = int(next(iterator))
            v_edge = int(next(iterator))
            edges.append([u_edge, v_edge])
        u = int(next(iterator))
        v = int(next(iterator))
            
        print(blocking_tokens(n, edges, u, v))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
